R/dataset-stream.R

```diff
@@ -0,0 +1,59 @@
+#' Tools for working with streaming data
+#'
+#' Crunch allows you to stream data to a dataset. Streaming data is useful for
+#' datasets which have frequent updates (see 
+#' [the Crunch documentation](http://docs.crunch.io/#streaming-rows) for more)
+#' information. `pendingMessages()` retrieves the number of pending messages and `appendStreamedRows()` appends all pending streamed rows to the dataset.
+#'
+#' @param x a CrunchDataset
+#' @param ... Additional batch attributes to set when appending pending stream 
+#' data. Options include
+#' `name` and `description`.
+#' @return number of pending messages for `pendingMessages()` and the dataset 
+#' with appended messages `appendStreamedRows()`
+#' @name streaming
+#' @aliases streaming pendingMessages appendStreamedRows
+#' @examples
+#' \dontrun{
+#' # need examples!
+#' }
+NULL
+
+
+#' @rdname streaming
+#' @export
+setMethod("pendingMessages", "CrunchDataset", function (x) {
+    stream_cat <- ShojiEntity(crGET(shojiURL(x, "fragments", "stream")))
+    stream_cat$pending_messages
+})
+
+#' Stream data to a Crunch dataset
+#' 
+#' @param ds a CrunchDataset
+#' @param data a dataframe with data to send as a stream, The given data values
+#'  must be in the Crunch I/O format (for example, category ids instead of 
+#'  names or numeric_values)
+#' @keywords internal
+streamRows <- function (ds, data) {
+    payload <- by(data, 1:nrow(data), function(row) toJSON(row) )
```

Comment:
The API docs say it should be one row per line. I've tried with lists like`{"var1":[1,2,3], "var2":["a", "b", "c"]}`, and when I try to append data streamed I get an error about decoding: `Cannot decode untyped 'list' value to class 'datetime'. (for variable 000004)` So, I do think this has to be JSONified by row. I've fixed the other problems (including not posting anything if there are no rows since there's no reason to).